subprojects/language-native/src/testFixtures/groovy/org/gradle/language/swift/AbstractSwiftComponentIntegrationTest.groovy

```diff
@@ -40,4 +43,93 @@ abstract class AbstractSwiftComponentIntegrationTest extends AbstractNativeLangu
         expect:
         succeeds "verifyBinariesSwiftVersion"
     }
+
+    def "throws exception when modifying Swift component language support after the binaries are known"() {
+        Assume.assumeThat(AbstractNativeLanguageComponentIntegrationTest.toolChain.version.major, Matchers.equalTo(4))
+
+        given:
+        makeSingleProject()
+        buildFile << """
+            ${componentUnderTestDsl} {
+                swiftLanguageVersionSupport = SwiftLanguageVersion.SWIFT3
+            }
+
+            ${componentUnderTestDsl}.binaries.whenElementKnown {
+                ${componentUnderTestDsl}.swiftLanguageVersionSupport = SwiftLanguageVersion.SWIFT4
+            }
+        """
+        settingsFile << "rootProject.name = 'swift-project'"
+
+        expect:
+        fails "verifyBinariesSwiftVersion"
+        failure.assertHasDescription("A problem occurred configuring root project 'swift-project'.")
+        failure.assertHasCause("This property is locked and cannot be changed.")
+    }
+
+    def "can build Swift 3 source code on Swift 4 compiler"() {
```

Comment:
We should also verify we can build Swift 3 with a swift 3 compiler and Swift 4 with a swift 4 compiler when the version is explicitly requested, and also what happens when no version is specified. We don't necessarily need to add a new test for this, as there are some test cases that implicitly cover this already.

It would be good if the test app could also probe the compiler or swift version and report on it, like the `CppCompilerDetectingTestApp`.

Reply:
Very good idea. Let me see what I can do.